src/crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.cc

```diff
@@ -0,0 +1,619 @@
+                                     std::nullopt, std::nullopt));
+        }
+        case mutation_status_t::WAS_SPLIT:
+          return handle_split(oc, child_pt, mresult);
+        default:
+          return rm_key_ertr::make_ready_future<mutation_result_t>(mresult);
+      }
+    });
+  });
+}
+
+OMapInnerNode::list_keys_ret
+OMapInnerNode::list_keys(omap_context_t oc, std::string &start, size_t max_result_size)
+{
+  logger().debug("OMapInnerNode: {}", __func__);
+  auto  child_iter = get_containing_child(start);
+
+  return seastar::do_with(child_iter, iter_end(), list_keys_result_t(), [=, &start]
+    (auto &biter, auto &eiter, auto &result) {
+    result.next = start;
+    return crimson::do_until([=, &biter, &eiter, &result] ()
+       -> list_keys_ertr::future<bool> {
+      if (biter == eiter  || result.keys.size() == max_result_size)
+        return list_keys_ertr::make_ready_future<bool>(true);
+
+      auto laddr = biter->get_node_key().laddr;
+      return omap_load_extent(oc, laddr, get_meta().depth - 1).safe_then(
+        [=, &biter, &eiter, &result] (auto &&extent) {
+        return extent->list_keys(oc, result.next, max_result_size - result.keys.size())
+          .safe_then([&biter, &eiter, &result] (auto &&list) mutable {
+          if (!list.keys.empty())
+            result.keys.insert(result.keys.end(), list.keys.begin(),list.keys.end());
+
+          biter++;
+          if (list.next != "")
+            result.next = list.next;
+          else if (biter != eiter)
+            result.next = biter->get_node_val();
+          else
+            result.next = "";
+
+          return list_keys_ertr::make_ready_future<bool>(false);
+        });
+      });
+    }).safe_then([&result, ref = OMapNodeRef(this)] {
+      return list_keys_ertr::make_ready_future<list_keys_result_t>(std::move(result));
+    });
+  });
+}
+
+OMapInnerNode::list_ret
+OMapInnerNode::list(omap_context_t oc, std::string &start, size_t max_result_size)
+{
+  logger().debug("OMapInnerNode: {}", __func__);
+  auto child_iter = get_containing_child(start);
+
+  return seastar::do_with(child_iter, iter_end(), list_kvs_result_t(), [=, &start]
+    (auto &biter, auto &eiter, auto &result) {
+    result.next = start;
+    return crimson::do_until([=, &biter, &eiter, &result] ()
+      -> list_ertr::future<bool> {
+      if (biter == eiter  || result.kvs.size() == max_result_size)
+        return list_ertr::make_ready_future<bool>(true);
+
+      auto laddr = biter->get_node_key().laddr;
+      return omap_load_extent(oc, laddr, get_meta().depth - 1).safe_then(
+        [=, &biter, &eiter, &result] (auto &&extent) {
+        return extent->list(oc, result.next, max_result_size - result.kvs.size())
+          .safe_then([&biter, &eiter, &result] (auto &&list) mutable {
+          if (!list.kvs.empty())
+            result.kvs.insert(result.kvs.end(), list.kvs.begin(),list.kvs.end());
+
+          biter++;
+          if (list.next != "")
```

Comment:
So "" is a special value of next that indicates that we've scanned everything in *this?  That would make a key equal to ""  behave incorrectly.  I think you should change std::string next to std::optional<std::string> next in list_keys_result_t.

Reply:
@athanatos : here  next set to be "" means search from iter_begin() in next node, there is no confliction with a key equal to "",  if I set std::string next to std::optional<std::string> next I need handle search functions  such as contains, get_containing_child, string_lower_bound etc to handle std::nullopt, and all other functions to call these search functions need  use std::optional<std::string> as input. it seems not make sense.  What's your opinion?

Reply:
if list_keys_result_t return with next = "", means we have already return all keys in omap tree.

Reply:
@athanatos:  here set next "" means we've scanned everything in this node.  and if it not get the max_result_size, the parent node will set next to next iter key(biter->get_node_val()), so will search next node from iter_begin(), (even it is "", also search from iter_begin()).  no confliction with a key equal to "".  If I change std::string next to std::optional<std::string> next in list_keys_result_t. I need change search functions such as contains, get_containing_child and string_lower_bounds etc to handle std::optional, the other functions calling these search functions also need handle std::optional. it seems don't make sense.  What's your opinion?

Reply:
@athanatos I have made a little change for the logic, please take a look.

Reply:
I think it's rather unintuitive for next to be set to a value that sorts *less* than every string in the event that we've finished scanning the node.

Reply:
sorry to comment out of nowhere, but treating empty string literal as a magic value seems suspect to me too

Reply:
I think it would be better to use an optional.  Also, add a comment to the top level interface explaining the meaning of the value.

Reply:
@athanatos change next to std::optional.